src/azure-cli-core/azure/cli/core/tests/test_command_registration.py

```diff
@@ -298,10 +298,10 @@ def _check_index():
         loader.load_command_table(["hello", "mod-only"])
         _check_index()
 
-        with mock.patch("azure.cli.core.__version__", "2.7.0"), mock.patch.object(cli.cloud, "profile", "2019-03-01-hybrid"):
```

Comment:
This test fails in Python 3.11:

```
E           AttributeError: <MagicMock id='2148541021584'> does not have the attribute '__version__'

C:\Users\xxx\AppData\Local\Programs\Python\Python311\Lib\unittest\mock.py:1416: AttributeError
```

I found a similar issue https://github.com/scikit-hep/pyhf/issues/2143. Switching to `mock.patch.object(azure.cli.core, '__version__',  "2.7.0")` indeed solves this error. This makes me wondering if the import logic with string `'azure.cli.core'` has been broken in Python 3.11.

After further debugging `mock.py` source code, I noticed `unittest.mock._patch.getter` in Python 3.10 is able to get `azure.cli.core` module, but `unittest.mock._patch.getter` in Python 3.11 is not able to do that - it returns a `MagicMock`. According to https://docs.python.org/3/library/unittest.mock.html#id3:

> The only exceptions are magic methods and attributes (those that have leading and trailing double underscores). Mock doesn’t create these but instead raises an [AttributeError](https://docs.python.org/3/library/exceptions.html#AttributeError). This is because the interpreter will often implicitly request these methods, and gets very confused to get a new Mock object when it expects a magic method. If you need magic method support see [magic methods](https://docs.python.org/3/library/unittest.mock.html#magic-methods).

Git blaming `mock.py` shows `unittest.mock._get_target`'s logic has been changed to use `pkgutil.resolve_name` in Python 3.11 (https://github.com/python/cpython/pull/18544), but `pkgutil.resolve_name`'s functionality is broken by

https://github.com/Azure/azure-cli/blob/87004321ec280e092fc41febf4ada3b58f01f388/src/azure-cli-core/azure/cli/core/tests/test_command_registration.py#L264

Reply:
Great work, I have never considered that this error is related to `@mock.patch('pkgutil.iter_modules', _mock_iter_modules) `.
Seems that it break all nested mock? I still have some other questions:

1. Why do we need this mock?
2. If we don't have this mock, will it fail? I see `__version__` is not in [support list](https://docs.python.org/3/library/unittest.mock.html#magic-methods).

Reply:
I figured this out.

The actual cause is `@mock.patch('importlib.import_module', _mock_import_lib)`.

In 3.10, mock use its own `_import(target_name)`, which uses `__import__` and `getattr`, to get the target object. `azure.cli.core` is returned.

In 3.11, it uses `pkgutil.resolve_name(target_name)` to get target object, which calls `importlib.import_module` internally.
However, `importlib.import_module` is patched and always returns `mock.MagicMock()`, thus a `MagicMock()` is returned and it does not have `__version__` attribution.

Reply:
Let's keep this conversation open as a TODO item. I feel there should be a way to patch `pkgutil.resolve_name` more elegantly.